### PR TITLE
Feature/plans

### PR DESCRIPTION
**Pull request by MoeAlshidi**
**Highlights**
- Personal Trainer package ( Repo, Service, Controller, Model)<MVC>.
- Introduced DTOs and refactored some code.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,7 @@ target/
 !**/src/main/**/target/
 !**/src/test/**/target/
 application.yml
+docker-compose.yml
```
Comment (amigoscode) [inline .gitignore:7]: this file should be checked in for other developers

File: pom.xml [hunk 2]
```diff
@@ -45,11 +56,20 @@
...
 			<scope>test</scope>
 		</dependency>
+
 		<dependency>
 			<groupId>org.springframework.boot</groupId>
 			<artifactId>spring-boot-starter-data-jpa</artifactId>
 		</dependency>
-	</dependencies>
+
+        <dependency>
```
Comment (amigoscode) [inline pom.xml:54]: what is this for?

File: src/main/java/com/project/trackfit/Main.java [hunk 3]
```diff
@@ -3,10 +3,12 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
+import java.time.LocalTime;
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/Main.java]: remove this

File: src/main/java/com/project/trackfit/core/model/CustomResponse.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+@JsonInclude(NON_NULL)
+public class CustomResponse {
+    protected LocalDateTime timeStamp;
+    protected  int statusCode;
+    protected HttpStatus status;
+    protected  String reason;
+    protected  String message;
+    protected String developerMessage;
+
+    protected Map<?,?> data;
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/core/model/CustomResponse.java]: never use maps

File: src/main/java/com/project/trackfit/core/model/CustomResponse.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.Map;
+
+import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
+
+@Data
+@SuperBuilder
+@JsonInclude(NON_NULL)
+public class CustomResponse {
+    protected LocalDateTime timeStamp;
+    protected  int statusCode;
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/core/model/CustomResponse.java]: 2 spaces between `protected` and `int`

File: src/main/java/com/project/trackfit/core/model/CustomResponse.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    protected HttpStatus status;
+    protected  String reason;
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/core/model/CustomResponse.java]: same here

File: src/main/java/com/project/trackfit/core/model/CustomResponse.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    protected  String message;
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/core/model/CustomResponse.java]: same here

File: src/main/java/com/project/trackfit/core/model/CustomResponse.java (hunk 4 above)
Comment (amigoscode) [reply src/main/java/com/project/trackfit/core/model/CustomResponse.java]: what is data?

File: src/main/java/com/project/trackfit/core/model/CustomResponse.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+import java.time.LocalDateTime;
+import java.util.Map;
+
+import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
+
+@Data
+@SuperBuilder
+@JsonInclude(NON_NULL)
+public class CustomResponse {
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/core/model/CustomResponse.java]: CustomResponse means nothing

File: src/main/java/com/project/trackfit/core/registration/EmailValidator.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.stereotype.Component;
+
+import java.util.regex.Pattern;
+
+@Component
+@PropertySource("classpath:application.properties")
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/core/registration/EmailValidator.java]: no need for this

File: src/main/java/com/project/trackfit/customer/CustomerController.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+public class CustomerController {
+
+    private final CustomerService customerService;
+
+    /**
+     * Spring Boot REST API creates a User
+     * http://[::1]:8080/api/v1/users/
+     */
+    @PostMapping
+    public ResponseEntity<CustomResponse> createUser(@Valid @RequestBody Customer customer){
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/customer/CustomerController.java]: just return the savedUserId.getCustomer_id data type

File: src/main/java/com/project/trackfit/customer/CustomerController.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        Customer savedUserId = customerService.createCustomer(customer);
+        return  ResponseEntity.ok(
+                CustomResponse.builder()
+                        .timeStamp(now())
+                        .data(Map.of("customer_id",savedUserId.getCustomer_id()))
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/customer/CustomerController.java]: getCustomerId

File: src/main/java/com/project/trackfit/customer/CustomerRetrieveRequestMapper.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
+package com.project.trackfit.customer;
+
+import org.springframework.stereotype.Service;
+
+import java.util.function.Function;
+@Service
+public class CustomerRetrieveRequestMapper  implements Function<Customer,RetrieveCustomerRequest> {
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/customer/CustomerRetrieveRequestMapper.java:7]: format code

File: src/main/java/com/project/trackfit/customer/RetrieveCustomerRequest.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package com.project.trackfit.customer;
+
+public record RetrieveCustomerRequest(
+        Long customer_id,
+        String firstName,
+        String LastName,
+        String email,
+        Integer age
+
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/customer/RetrieveCustomerRequest.java]: remove multiline spaces

File: src/main/java/com/project/trackfit/customer/RetrieveCustomerRequest.java [hunk 14]
```diff
@@ -0,0 +1,12 @@
+package com.project.trackfit.customer;
+
+public record RetrieveCustomerRequest(
+        Long customer_id,
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/customer/RetrieveCustomerRequest.java]: customerId

File: src/main/java/com/project/trackfit/subscription/CreateSubscriptionRequest.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+import com.project.trackfit.customer.Customer;
+import com.project.trackfit.trainer.PersonalTrainer;
+
+import java.time.LocalDateTime;
+
+
+import java.util.UUID;
+
+public record CreateSubscriptionRequest(
+
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/subscription/CreateSubscriptionRequest.java]: remove this line

File: src/main/java/com/project/trackfit/subscription/CreateSubscriptionRequest.java [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+         UUID personalTrainerId,
+         Long customerId,
+         LocalDateTime subscribedAt
+
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/subscription/CreateSubscriptionRequest.java]: same here

File: src/main/java/com/project/trackfit/subscription/RetrieveSubscriptionRequest.java [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.UUID;
+
+public record RetrieveSubscriptionRequest(
+         UUID id,
+         LocalDateTime subscribedAt,
+         LocalDateTime expiredOn,
+         Boolean isActive,
+         RetrieveTrainerRequest personalTrainer,
+         RetrieveCustomerRequest customer
+
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/subscription/RetrieveSubscriptionRequest.java]: here too

File: src/main/java/com/project/trackfit/subscription/SubscriptionController.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    @PostMapping
+    public ResponseEntity<CustomResponse>subscribe(@Valid @RequestBody CreateSubscriptionRequest subscriptionRequest){
+
+        subscriptionService.createSubscription(subscriptionRequest);
+
+        return ResponseEntity.ok(
+                CustomResponse.builder()
+                        .timeStamp(now())
+                        .data(Map.of("Subscription",""))
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/subscription/SubscriptionController.java]: why return empty value for map

File: src/main/java/com/project/trackfit/subscription/SubscriptionRetrieveRequestMapper.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.function.Function;
+
+@Service
+@AllArgsConstructor
+public class SubscriptionRetrieveRequestMapper implements Function<Subscription,RetrieveSubscriptionRequest> {
+    private final TrainerRetrieveRequestMapper trainerRetrieveRequestMapper;
+    private  final CustomerRetrieveRequestMapper customerRetrieveRequestMapper;
+    @Override
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/subscription/SubscriptionRetrieveRequestMapper.java:19]: new line

File: src/main/java/com/project/trackfit/subscription/SubscriptionService.java [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+
+
+import java.util.UUID;
+
+public interface SubscriptionService {
+    void createSubscription(CreateSubscriptionRequest subscription);
+    Subscription findSubscriptionByID(UUID subscription_id);
+    Iterable<RetrieveSubscriptionRequest> findAllSubscription();
+
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/subscription/SubscriptionService.java:12]: remove line

File: src/main/java/com/project/trackfit/trainer/PersonalTrainer.java [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.util.Set;
+import java.util.UUID;
+
+@Entity
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class,property = "id")
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/trainer/PersonalTrainer.java]: do you need this?

File: src/main/java/com/project/trackfit/trainer/PersonalTrainerController.java [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+
+@RestController
+@AllArgsConstructor
+@RequestMapping("api/v1/trainers")
+public class PersonalTrainerController {
+    private  final PersonalTrainerService personalTrainerService;
+    private  final ModelMapper modelMapper;
+
+    private RetrieveTrainerRequest convertToDto(PersonalTrainer entity)
+    {
```
Comment (amigoscode) [inline src/main/java/com/project/trackfit/trainer/PersonalTrainerController.java]: this is not java format
```java
private RetrieveTrainerRequest convertToDto(PersonalTrainer entity) 
{
}
```

```
private RetrieveTrainerRequest convertToDto(PersonalTrainer entity) {

}
```

File: src/main/resources/application.properties [hunk 23]
```diff
@@ -2,6 +2,9 @@ spring.datasource.url=jdbc:mysql://localhost:3306/fitness_tracker
...
...\\t])*))*(?:,@(?:(?:\\r\\n)?[ \\t])*(?:[^()<>@,;:\\\\\".\\[\\] \\000-\\031]+(?:(?:(?:\\r\\n)?[ \\t])+|\\Z|(?=[\\[\"()<>@,;:\\\\\".\\[\\]]))|\\[([^\\[\\]\\r\\\\]|\\\\.)*\\](?:(?:\\r\\n)?[ \\t])*)(?:\\.(?:(?:\\r\\n)?[ \\t])*(?:[^()<>@,;:\\\\\".\\[\\] \\000-\\031]+(?:(?:(?:\\r\\n)?[ \\t])+|\\Z|(?=[\\[\"()<>@,;:\\\\\".\\[\\]]))|\\[([^\\[\\]\\r\\\\]|\\\\.)*\\](?:(?:\\r\\n)?[ \\t])*))*)*:(?:(?:\\r\\n)?[ \\t])*)?(?:[^()<>@,;:\\\\\".\\[\\] \\000-\\031]+(?:(?:(?:\\r\\n)?[ \\t])+|\\Z|(?=[\\[\"()<>@,;:\\\\\".\\[\\]]))|\"(?:[^\\\"\\r\\\\]|\\\\.|(?:(?:\\r\\n)?[ \\t]))*\"(?:(?:\\r\\n)?[ \\t])*)(?:\\.(?:(?:\\r\\n)?[ \\t])*(?:[^()<>@,;:\\\\\".\\[\\] \\000-\\031]+(?:(?:(?:\\r\\n)?[ \\t])+|\\Z|(?=[\\[\"()<>@,;:\\\\\".\\[\\]]))|\"(?:[^\\\"\\r\\\\]|\\\\.|(?:(?:\\r\\n)?[ \\t]))*\"(?:(?:\\r\\n)?[ \\t])*))*@(?:(?:\\r\\n)?[ \\t])*(?:[^()<>@,;:\\\\\".\\[\\] \\000-\\031]+(?:(?:(?:\\r\\n)?[ \\t])+|\\Z|(?=[\\[\"()<>@,;:\\\\\".\\[\\]]))|\\[([^\\[\\]\\r\\\\]|\\\\.)*\\](?:(?:\\r\\n)?[ \\t])*)(?:\\.(?:(?:\\r\\n)?[ \\t])*(?:[^()<>@,;:\\\\\".\\[\\] \\000-\\031]+(?:(?:(?:\\r\\n)?[ \\t])+|\\Z|(?=[\\[\"()<>@,;:\\\\\".\\[\\]]))|\\[([^\\[\\]\\r\\\\]|\\\\.)*\\](?:(?:\\r\\n)?[ \\t])*))*\\>(?:(?:\\r\\n)?[ \\t])*))*)?;\\s*)
```
Comment (amigoscode) [inline src/main/resources/application.properties]: what is this?

**Review by ghost — DISMISSED** [review]

File: src/main/java/com/project/trackfit/core/model/CustomResponse.java (hunk 8 above)
Comment (MoeAlshidi) [reply src/main/java/com/project/trackfit/core/model/CustomResponse.java]: Changed it to APICustomResponse

**Review by Andreas-Kreouzos — APPROVED** [review]

**Pull request merged**